File: lib/service/wallet_service.rb
require_relative '../model/types/result'
require_relative '../model/wallet'

def print_wallet(wallet)
  puts "Wallet:"
  puts "  sold: #{wallet.sold}"
end

def deposit(wallet)
  print "Money to deposit: "
  input_str = gets.chomp

  begin
    input_int = Integer(input_str)
    wallet.sold = input_int
    Result.new(true, "Deposit successfully", wallet)
  rescue ArgumentError, TypeError
    Result.new(false, "Invalid Input - retry", input_str)
  end
end

def withdrawal(wallet)
  print "Amount to withdraw: "
  input_str = gets.chomp

  begin
    input_int = Integer(input_str)
    if input_int > wallet.sold
      Result.new(false, "Insufficient balance", input_int)
    else
      wallet.sold -= input_int
      Result.new(true, "Withdrawal successfully", wallet)
    end
  rescue ArgumentError, TypeError
    Result.new(false, "Invalid Input - retry", input_str)
  end
end

File: src/service/wallet_service.py
from src.model.type.result import Result
from src.model.wallet import Wallet


def print_wallet(wallet: Wallet) -> None:
    print(f"Wallet:\n  sold: {wallet.sold}")


def deposit(wallet: Wallet) -> Result:
    input_str: str = input("Money to deposit: ")
    try:
        input_int: int = int(input_str)
        wallet.sold = input_int
        return Result(True, "Deposit successfully", wallet)
    except ValueError:
        return Result(False, "Invalid Input - retry", input_str)


def withdrawal(wallet: Wallet) -> Result:
    input_str: str = input("Amount to withdraw: ")
    try:
        input_int: int = int(input_str)
        if input_int > wallet.sold:
            return Result(False, "Insufficient balance", input_int)
        wallet.sold = wallet.sold - input_int
        return Result(True, "Withdrawal successfully", wallet)
    except ValueError:
        return Result(False, "Invalid Input - retry", input_str)
